ltq-dsl-base: use ActualNetDataRate for DSL bitrate statistics

jekkos opened the pull request:
Both my ISP and Fr!tzbox 7490 consistently report slightly different speeds then OpenWrt. Before this patch, my ISP reports the same speed as the 7490 reports, which was always different from what OpenWrt reported.

This patch fixes this discrepancy. 

Signed-off-by: Jeroen Peelaerts <jeroen.peelaerts@gmail.com>

adschm commented:
This changes ltq-dsl-base package, so commit title prefix should be "ltq-dsl-base:".

Apart from that, your commit title is apparently too long.

adschm commented:
Please also bump PKG_RELEASE.

jekkos commented:
Ok I have corrected the commit message and bumped package version.

dhewg commented:
Just stumbled upon the fact that the older danube dsl support doesn't have this field, only `ActualDataRate`.
See `DSL_G997_ChannelStatusData_t` in `drv_dsl_cpe_api-3.24.4.4/src/include/drv_dsl_cpe_api_g997.h`

So this would break on those boards, I don't know if they're still relevant though. Maybe @xdarklight can assist?

xdarklight commented:
> Just stumbled upon the fact that the older danube dsl support doesn't have this field, only `ActualDataRate`.
> See `DSL_G997_ChannelStatusData_t` in `drv_dsl_cpe_api-3.24.4.4/src/include/drv_dsl_cpe_api_g997.h`
> 
> So this would break on those boards, I don't know if they're still relevant though. Maybe @xdarklight can assist?

I think I have a board with Danube SoC somewhere
But I cannot help with (runtime) testing anything that requires a working ADSL connection, where I live it's now VDSL-only

dhewg commented:
I meant how to proceed, does `ActualNetDataRate` make more sense than `ActualDataRate`?

Digging around... from `drv_dsl_cpe_api-4.17.18.6/src/include/drv_dsl_cpe_api_g997.h`:
```
   /**
   Actual Data Rate.

   If retransmission is not used in a given transmit direction:
   this parameter reports the actual net data rate at which the bearer channel
   is operating (in L2, the actual L2 net data rate is reported).

   If retransmission is used in a given transmit direction:
   this parameter reports the Expected Throughput (ETR) (as defined in G.inp)
   at which the bearer channel is operating.

   If retransmsission is used in a given direction: the reporting of
   ActualDataRate is according to  ITU-T G.997.1 chapter 7.5.2.1.
   If retransmsission is not used in a given direction: the reporting of
   ActualDataRate is not according to  ITU-T G.997.1 chapter 7.5.2.1
   during L2 power state but the reporting of parameter ActualNetDataRate
   is according to  ITU-T G.997.1 chapter 7.5.2.1 during all power states.

   The data rate is coded in bit/s.
   According to G992.1 Table 6-3:
   "Net data rate" + Frame overhead rate = "Aggregate data rate"
   "Aggregate data rate" + RS Coding overhead rate = "Total data rate"
   "Total data rate" + Trellis Coding overhead rate = Line rate */
   DSL_OUT DSL_uint32_t ActualDataRate;
```
and
```
   /**
   Actual Net Data Rate (ACTNDR)

   Refer to ITU-T G.997.1 chapter 7.5.2.8

   Independent whether retransmission is used or not in a given transmit
   direction:
   -   In L0 state, this parameter reports the Net Data Rate (as specified
   in G.992.3, G.992.5 or G.993.2) at which the bearer channel is operating.
   -   In  L2 state, the parameter contains the Net Data Rate (as specified
   in G.992.3, G.992.5 or G.993.2) in the previous L0 state.

   The data rate is coded in bit/s. */
   DSL_OUT DSL_uint32_t ActualNetDataRate;
```

That sounds like we should use `ActualNetDataRate` since  it's independent from retx/g.inp. Since the ancient adsl driver doesn't support it, for my recent patch set that would translate to:
```
diff --git a/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c b/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c
index 9f73453ef8..434d5645b5 100644
--- a/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c
+++ b/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c
@@ -532,7 +532,11 @@ static void g997_channel_status(int fd, DSL_AccessDir_t direction) {
    IOCTL_DIR(DSL_G997_ChannelStatus_t, DSL_FIO_G997_CHANNEL_STATUS_GET, direction);
 
    m_u32("interleave_delay", out.data.ActualInterleaveDelay * 10);
+#ifndef INCLUDE_DSL_CPE_API_DANUBE
+   m_u32("data_rate", out.data.ActualNetDataRate);
+#else
    m_u32("data_rate", out.data.ActualDataRate);
+#endif
 }
 
 static void g997_line_status(int fd, DSL_AccessDir_t direction) {
```

Does that make sense?

jekkos commented:
Seems like a good solution indeed @dhewg .  So indeed my observation aligns with how this statistic is documented, good to see. I guess that your improvements will superseed this patch.

dhewg commented:
Let's keep this PR open though. I don't want to 'sneak in' a functional change in my set. That's meant to just switch to ubus.
If/when my set gets merged, you can then adapt your patch with my diff above (and keep the credit ;). A distinct commit is more clear and easier to bisect/revert

dhewg commented:
This can now be rebased. I've been running this for weeks now:

```
diff --git a/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c b/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c
index 52b2be20e1..8eefdfe4d5 100644
--- a/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c
+++ b/package/network/config/ltq-vdsl-app/src/src/dsl_cpe_ubus.c
@@ -532,7 +532,12 @@ static void g997_channel_status(int fd, DSL_AccessDir_t direction) {
    IOCTL_DIR(DSL_G997_ChannelStatus_t, DSL_FIO_G997_CHANNEL_STATUS_GET, direction);
 
    m_u32("interleave_delay", out.data.ActualInterleaveDelay * 10);
+#ifndef INCLUDE_DSL_CPE_API_DANUBE
+   // prefer ACTNDR, see comments in drv_dsl_cpe_api_g997.h
+   m_u32("data_rate", out.data.ActualNetDataRate);
+#else
    m_u32("data_rate", out.data.ActualDataRate);
+#endif
 }
 
 static void g997_line_status(int fd, DSL_AccessDir_t direction) {
```

adschm commented:
@dhewg FYI: You might speed these other PRs up by providing Reviewed-by and/or Tested-by once they are updated.

dhewg commented:
That was the idea, I just didn't want to take the credit with a patch with my name on it :)

jekkos commented:
Ok great to see your changes were merged. I'll rebase tonight.

jekkos committed:
lantiq: use ActualNetDataRate for speed reporting

Switch to Actual Net Data Rate (ACTNDR) for speed reporting on lantiq VDSL modems

Refer to ITU-T G.997.1 chapter 7.5.2.8

Independent whether retransmission is used or not in a given transmit direction:
-   In L0 state, this parameter reports the Net Data Rate (as specified in G.992.3, G.992.5 or G.993.2) at which the bearer channel is operating.
-   In  L2 state, the parameter contains the Net Data Rate (as specified in G.992.3, G.992.5 or G.993.2) in the previous L0 state.

Signed-off-by: Jeroen Peelaerts <jeroen.peelaerts@gmail.com>
Reviewed-by: Andre Heider <a.heider@gmail.com>

dhewg commented:
Having said that, if you just use the chunk I posted above you can already add:
`Reviewed-by: Andre Heider <a.heider@gmail.com>`

jekkos commented:
> Having said that, if you just use the chunk I posted above you can already add:
> `Reviewed-by: Andre Heider <a.heider@gmail.com>`

Should be good now. I Added the Reviewed-by.

dhewg reviewed (approved):
Maybe expand the commit message? I think the key point is that this is independent of ReTx?
```
   Actual Net Data Rate (ACTNDR)

   Refer to ITU-T G.997.1 chapter 7.5.2.8

   Independent whether retransmission is used or not in a given transmit
   direction:
   -   In L0 state, this parameter reports the Net Data Rate (as specified
   in G.992.3, G.992.5 or G.993.2) at which the bearer channel is operating.
   -   In  L2 state, the parameter contains the Net Data Rate (as specified
   in G.992.3, G.992.5 or G.993.2) in the previous L0 state.
```

adschm commented:
Title is too long and commit message is missing.

jekkos commented:
@adschm Bear with me, first time contribution. I've added some explanation in the commit + made the title shorter.

dhewg commented:
@hauke I think this is good to go, please throw this on your can-be-merged-pile ;)

dhewg commented:
Ping, anyone?

jekkos commented:
@adschm I have addressed most comments, think this change is ready to go?

hauke commented:
Thank you for the pull request, I applied it to master.
